### ISSUE TITLE
Setting @ErrorHandler doesn't catch SocketTimeoutException

### ISSUE THREAD

**Issue by kenkyee** [yongjhih/retrofit#6]
Tested by adding the @ErrorHandler annotation before my class:

```
@Retrofit.ErrorHandler(MYErrorHandler.class)
public abstract class TestCreationApi {
```

My handler looks like the example but it never gets called:

```
public class MyErrorHandler implements ErrorHandler {
    @Override
    public Throwable handleError(RetrofitError cause) {
        Response r = cause.getResponse();
        if (r != null && r.getStatus() == 401) {
            // clear authToken because we've been logged out by the server
            AuthApiManager.setAuthToken(null);

            Log.d("MyErrorHandler", "Logged out because of bad token");

            return new RuntimeException("401", cause);
        }
        return cause;
    }
}
```

Breakpoints don't get hit in the code and the print statement doesn't seem to work.  Generated code seem to set it though.

```
        if (this.errorHandler == null) {
            this.errorHandler = new com.serverapi.MYErrorHandler();
        }
```

It doesn't catch the 401 error or a sockettimeout.  You can reproduce the sockettimeout by disconnecting your internet network temporarily on the emulator (genymotion) so that's easier to test than the 401.

**Comment by yongjhih** [comment]
It should be fixed: b2b66075fd

**Comment by kenkyee** [comment]
Doesn’t seem to work still…when I shut off wifi, the error handler’s handleError() never gets called…an exception happens here instead...

**Comment by kenkyee** [comment]
Forgot to mention I did add this to my gradle file:

compile 'com.github.yongjhih.NotRetrofit:retrofit:b2b6607'
apt 'com.github.yongjhih.NotRetrofit:retrofit-processor:b2b6607'

**Comment by kenkyee** [comment]
![screen shot 2015-11-12 at 4 59 45 pm](https://cloud.githubusercontent.com/assets/12235883/11132686/5c333016-895f-11e5-8a42-a0d762e5c159.png)

**Comment by yongjhih** [comment]
Hi Ken,

Sorry for replying late, I was on holiday.

It should be fixed by 06d6be7.

**Comment by kenkyee** [comment]
Thanks.  Hope you had a good holiday :-)

FWIW, I tried pinning to the latest:
compile 'com.github.yongjhih.NotRetrofit:retrofit:06d6be7'
apt 'com.github.yongjhih.NotRetrofit:retrofit-processor:06d6be7'

but my code in my error handler never runs.  I think it’s because it gets no exception when the server sends a 401…it sends JSON back w/ an error msgs so it doesn’t cause any exceptions in your code so the error handler never gets calls.  The error handler does get called if it can’t reach the server when I shut off wifi though, so I know the error handler is hooked in right.

 ken

**Comment by yongjhih** [comment]
You are right, I also found it. So i committed 6de8721: Support httpError exception throwing.

**Comment by kenkyee** [comment]
I’m not getting an error code.
I think I see a problem in your MainTest unit test.
You’re checking that the error handler gets called, in your mock service, but you’re not checking that it’s passing through the response code.  I see that when the error handler gets hit too…the response field is null.

With that checkin, I’m also finding that when I get an error response code, the JSON in the body of the response is no longer being parsed like it used to be.  I can work around this, but it’s a different behavior than before.  Unit tests are definitely a good idea :-)

Hope that helps.

 ken

**Comment by yongjhih** [comment]
It's mistake that did not check response code in `testHttpErrorHandler()`. But I had checked the testing log printed that represented the right response code of 404 before, not null. I think you need to clean build after updated version.

And welcome any patch for unit tests to show the problem up.

**Comment by kenkyee** [comment]
Cleaning build didn’t help.
I’ll try to write up a unit test for it and submit a PR :-)

 thanks,

 ken

**Comment by kenkyee** [comment]
Added unit tests and confirmed error handler gets called when using the MockWebServer.  The unit tests show the issue of not getting the response body if the error handler is called w/ an error response code.

Not sure why it’s not acting the same way on my system w/ a real web server but will investigate more.

**Comment by yongjhih** [comment]
Thanks for PR, it has been merged #8 and fixed `null` responseBody by f0f93c42
